cli/Makefile

```
@@ -54,13 +53,34 @@ $(BUILDDIR)/julia_res.o: $(JULIAHOME)/contrib/windows/julia.rc $(JULIAHOME)/VERS
 		'/^(\d+)\.?(\d*)\.?(\d*)/ && \
 		print int $$1,",",int $$2,",0,",int $$3'` && \
 	$(CROSS_COMPILE)windres $< -O coff -o $@ -DJLVER=$$JLVERi -DJLVER_STR=\\\"$$JLVER\\\"
-OBJS += julia_res.o
-DOBJS += julia_res.o
+OBJS += $(BUILDDIR)/julia_res.o
+DOBJS += $(BUILDDIR)/julia_res.o
 endif
 endif
 
-julia-release: $(build_bindir)/julia$(EXE) $(build_shlibdir)/libjulialoader.$(JL_MAJOR_MINOR_SHLIB_EXT)
-julia-debug: $(build_bindir)/julia-debug$(EXE) $(build_shlibdir)/libjulialoader-debug.$(JL_MAJOR_MINOR_SHLIB_EXT)
+ifneq ($(OS),WINNT)
+HIDDEN_SYMBOL := __attribute__((visibility(\"hidden\")))
+endif
+
+# Create a list of all `jl_` symbols that libjulia-internals contains that libjulia will re-export
+$(BUILDDIR)/jl_exports.h: $(SRCDIR)/../src/jl_exported_funcs.list $(SRCDIR)/../src/jl_exported_data.list
+	@echo "// This list of symbols auto-generated by cli/Makefile and sourced from src/jl_exported_*.list; do not edit" > $@
+	@echo "// Define Data symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_data.list | awk '{ print "JL_DLLEXPORT void * " $$1 ";" }' >> $@
+	@echo "// Function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "$(HIDDEN_SYMBOL) void * " $$1 "_addr;" }' >> $@
+	@echo >> $@
+	@echo "void loader_export_symbols(void * libjulia_internal) {" >> $@
+	@echo "    // Import function symbols" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "    " $$1 "_addr = dlsym(libjulia_internal, \"" $$1 "\");" }' >> $@
+	@echo "}" >> $@
+	@echo "// Define function trampolines" >> $@
+	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "JL_DLLEXPORT JL_NAKED void " $$1 "(void) { __asm__(\"jmpq *%0\" : : \"a\"(" $$1 "_addr) :); };" }' >> $@
```

Since you're going to need a different macro for every platform, I'd suggest splitting this to a decl and an architecture #define:
```suggestion
	@echo "#define JL_DLLREEXPORT(f) JL_DLLEXPORT JL_NAKED void #f(void) { __asm__(\"jmpq *%0\" : : \"a\"#f#_addr : ); };" }' >> $@
	@cat $(SRCDIR)/../src/jl_exported_funcs.list | awk '{ print "JL_DLLREEXPORT(" $$1 ")" }' >> $@
```

---

Ah, I ended up creating a `JL_JUMPTO(x)` and using that instead.  I think that should be equivalent?